src/objective/rank_xendcg_objective.hpp

```diff
@@ -0,0 +1,145 @@
+/*!
+ * Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#ifndef LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+#define LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+
+#include <LightGBM/objective_function.h>
+#include <LightGBM/utils/common.h>
+#include <LightGBM/utils/random.h>
+
+#include <string>
+#include <vector>
+
+namespace LightGBM {
+/*!
+* \brief Implementation of the learning-to-rank objective function, XE_NDCG [arxiv.org/abs/1911.09798].
+*/
+class RankXENDCG: public ObjectiveFunction {
+ public:
+  explicit RankXENDCG(const Config& config) {
+    if (config.seed != 0) {
+      rand_ = new Random(config.seed);
+    } else {
+      rand_ = new Random();
+    }
```

Comment:
@guolinke Does LightGBM require special treating for `seed == 0`? I thought seed is always required for reproducibility.

Reply:
If 'seed' is not set the expected behavior should be that a seed is selected at random, right? Here ideally I'd like to check if a seed has been set, though the only way to do that I believe is comparing with 0. Is there a different, preferred way of checking if seed is set/not-set?

Reply:
@sbruch 
> If 'seed' is not set the expected behavior should be that a seed is selected at random, right?

I don't think so.

When `seed` is set, it's used to generate other seeds:
https://github.com/microsoft/LightGBM/blob/f6b8ecf648f77f48d95a15e9ba5220e52f5d584c/src/io/config.cpp#L187-L195

Otherwise, default values of concrete seeds are used. For instance, 
https://github.com/microsoft/LightGBM/blob/f6b8ecf648f77f48d95a15e9ba5220e52f5d584c/include/LightGBM/config.h#L270
https://github.com/microsoft/LightGBM/blob/f6b8ecf648f77f48d95a15e9ba5220e52f5d584c/include/LightGBM/config.h#L290
https://github.com/microsoft/LightGBM/blob/f6b8ecf648f77f48d95a15e9ba5220e52f5d584c/include/LightGBM/config.h#L349

I guess, we need a new param, let say, `objective_seed` to be consistent with the existing codebase. I can't find any references where `seed` is used directly.

cc @guolinke

Reply:
I'm fine doing whatever you and @guolinke decide is appropriate. As far as I understand, either:

1) Reduce the highlighted code to: rand_ = new Random(config.seed); or,
2) Introduce an objective_seed

are what you have suggested so far.

As an aside, note that "new Random()" does not factor config.seed in the initialization at all and sets the seed to a randomly generated number. That's what gave me the impression that if seed is unset the behavior is expected to be non-deterministic.

Reply:
I personally vote for `#2` "Introduce an objective_seed". Let's wait for other opinions.

Reply:
Hi, I was wondering if you all have made a decision?

Reply:
I think an `objective_seed` is good.

Reply:
Done.